File: backend/handler/disk.go
package handler

import (
	"encoding/json"
	"net/http"

	"m3u8-downloader-web/model"
)

type DiskHandler struct{}

func NewDiskHandler() *DiskHandler {
	return &DiskHandler{}
}

type DiskInfo struct {
	MountPoint  string  `json:"mountPoint"`
	Device      string  `json:"device"`
	Total       uint64  `json:"total"`
	Used        uint64  `json:"used"`
	Free        uint64  `json:"free"`
	FreePercent float64 `json:"freePercent"`
	UsedPercent float64 `json:"usedPercent"`
	ColorClass  string  `json:"colorClass"`
	FsType      string  `json:"fsType"`
}

func (h *DiskHandler) GetDiskInfo(w http.ResponseWriter, r *http.Request) {
	path := "."
	
	if r.URL.Query().Get("path") != "" {
		path = r.URL.Query().Get("path")
	}
	
	diskInfo, err := getDiskInfo(path)
	if err != nil {
		h.sendError(w, http.StatusInternalServerError, "获取磁盘信息失败: "+err.Error())
		return
	}
	
	h.sendSuccess(w, diskInfo)
}

func (h *DiskHandler) GetAllDisks(w http.ResponseWriter, r *http.Request) {
	disks, err := getAllDisks()
	if err != nil {
		h.sendError(w, http.StatusInternalServerError, "获取磁盘列表失败: "+err.Error())
		return
	}
	
	h.sendSuccess(w, disks)
}

func (h *DiskHandler) CheckSpace(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Path string `json:"path"`
	}
	
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.sendError(w, http.StatusBadRequest, "无效的请求体")
		return
	}
	
	path := req.Path
	if path == "" {
		path = "."
	}
	
	diskInfo, err := getDiskInfo(path)
	if err != nil {
		h.sendError(w, http.StatusInternalServerError, "检查磁盘空间失败: "+err.Error())
		return
	}
	
	h.sendSuccess(w, diskInfo)
}

func getDiskInfo(path string) (*DiskInfo, error) {
	info, err := getDiskSpace(path)
	if err != nil {
		return nil, err
	}
	
	total := info.total
	free := info.free
	used := total - free
	
	freePercent := float64(free) / float64(total) * 100
	usedPercent := float64(used) / float64(total) * 100
	
	return &DiskInfo{
		MountPoint:  path,
		Total:       total,
		Used:        used,
		Free:        free,
		FreePercent: freePercent,
		UsedPercent: usedPercent,
		ColorClass:  getColorClass(usedPercent),
	}, nil
}

func getColorClass(usedPercent float64) string {
	if usedPercent >= 90 {
		return "danger"
	} else if usedPercent >= 75 {
		return "warning"
	} else if usedPercent >= 50 {
		return "info"
	}
	return "success"
}

type diskSpaceInfo struct {
	total uint64
	free  uint64
}

var getDiskSpace func(path string) (*diskSpaceInfo, error)
var getAllDisks func() ([]*DiskInfo, error)

func (h *DiskHandler) sendSuccess(w http.ResponseWriter, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(model.APIResponse{
		Code:    200,
		Message: "success",
		Data:    data,
	})
}

func (h *DiskHandler) sendError(w http.ResponseWriter, status int, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(model.APIResponse{
		Code:    status,
		Message: message,
	})
}

File: frontend/src/stores/disk.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { diskAPI } from '@/api'

export interface DiskInfo {
  mountPoint: string
  device: string
  total: number
  used: number
  free: number
  freePercent: number
  usedPercent: number
  colorClass: string
  fsType: string
}

export const useDiskStore = defineStore('disk', () => {
  const diskInfo = ref<DiskInfo | null>(null)
  const allDisks = ref<DiskInfo[]>([])
  const checkingSpace = ref(false)

  const getDiskInfo = async () => {
    try {
      const response = await diskAPI.getInfo()
      if (response.data.code === 200 && response.data.data) {
        diskInfo.value = response.data.data
      }
    } catch (error) {
      console.error('Failed to get disk info:', error)
    }
  }

  const getAllDisks = async () => {
    try {
      const response = await diskAPI.getAllDisks()
      if (response.data.code === 200 && response.data.data) {
        allDisks.value = response.data.data
      }
    } catch (error) {
      console.error('Failed to get all disks:', error)
    }
  }

  const checkSpace = async (path?: string, requiredMB?: number): Promise<{ enough: boolean; freeMB: number; requiredMB?: number }> => {
    checkingSpace.value = true
    try {
      const response = await diskAPI.checkSpace(path)
      if (response.data.code === 200 && response.data.data) {
        const freeMB = response.data.data.free / (1024 * 1024)
        if (requiredMB !== undefined) {
          return {
            enough: freeMB >= requiredMB,
            freeMB: Math.floor(freeMB),
            requiredMB
          }
        }
        return {
          enough: true,
          freeMB: Math.floor(freeMB)
        }
      }
      return { enough: true, freeMB: 0 }
    } catch (error) {
      console.error('Failed to check disk space:', error)
      return { enough: true, freeMB: 0 }
    } finally {
      checkingSpace.value = false
    }
  }

  const formatSize = (bytes: number): string => {
    if (bytes < 1024) return `${bytes} B`
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`
    if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
    return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`
  }

  const getColorClass = (colorClass: string): string => {
    switch (colorClass) {
      case 'danger':
        return 'bg-red-500'
      case 'warning':
        return 'bg-yellow-500'
      case 'info':
        return 'bg-blue-500'
      case 'success':
      default:
        return 'bg-green-500'
    }
  }

  const getStatusText = (colorClass: string): string => {
    switch (colorClass) {
      case 'danger':
        return '空间紧张'
      case 'warning':
        return '使用较高'
      case 'info':
        return '正常使用'
      case 'success':
      default:
        return '空间充足'
    }
  }

  return {
    diskInfo,
    allDisks,
    checkingSpace,
    getDiskInfo,
    getAllDisks,
    checkSpace,
    formatSize,
    getColorClass,
    getStatusText
  }
})